09-schema/03-rules.md

```diff
@@ -274,6 +274,112 @@ TypeQLDefine query = TypeQL.define(
 </div>
 
 This rule will make every relation transitive.
+## Efficient transitivity
+
+A common use-case for rules is to infer the transitive closure of a relation. The most straight-forward way of doing this is as follows:
+```typeql
+define
+
+rule transitive-reachability:
+when{
+    (from: $x, to: $y) isa edge;
+    (from: $y, to: $z) isa edge;
+} then {
+    (from: $x, to: $z) isa edge;
+};
+```
+We can interpret this rule as joining two paths together. In a chain `p-q-r-s-t`, to find all nodes reachable from p, we would generate the following relations:
+```
+p--q, q--r, r--s, s--t  (already in the database)
+
+p--r, q--s, r--t,       (Inferred)
+p--s, q--t, p--t        (Inferred)                    
+```
+
+Concretely, We would generate an `edge` relation _for every pair_ of nodes reachable from `p` - a **quadratic** number of relations.
+
+ The following section describes a recipe to answer forward transitivity queries materialising only a **linear** number of relations. Later, the recipe is extended to backward queries and undirected relations.
+
+### Forward transitivity
+We first define separate types for the persisted and (inferred) transitive version of the relation.
+For the example above, we use `edge` as the base relation type and `forward-reachable` as the inferred relation. We then update the rule as follows: 
+```typeql
+define
+rule forward-transitivity-base:
+when{
+    (from: $x, to: $y) isa edge;
+} then {
+    (from: $x, to: $y) isa forward-reachable;
+};
+
+rule forward-transitivity-recursive:
+when{
+    (from: $x, to: $y) isa forward-reachable;
+    (from: $y, to: $z) isa edge;
+} then {
+    (from: $x, to: $z) isa forward-reachable;
+};
+```
+
+We can intepret this as finding a path and extend it by one. Querying all nodes reachable from p in the chain p-q-r-s-t would generate the following relations:
```

Comment:
show the query you would run to trigger the rule, and definitely need to show that `from` is bound